File: routes/chats.py
"""Chat and message routes — /api/chats"""
from flask.views import MethodView
from flask_smorest import Blueprint as SmorestBlueprint, abort
from flask import request, current_app
from datetime import datetime
import uuid
from sqlalchemy import or_

from schemas import ChatObject, MessageObject, MessagesQueryArgs, SendMessageBody
from routes import require_auth
from services.users import get_profile


chats_blp = SmorestBlueprint("chats", __name__, url_prefix="/api/chats",
                              description="Chat conversations and messages")


# ── / (conversation list) ────────────────────────────────────────────────

@chats_blp.route("/")
class ChatListResource(MethodView):
    @chats_blp.response(200, ChatObject(many=True))
    def get(self):
        """List all conversations for the authenticated user"""
        require_auth()
        from app import Message, DOMAIN, Webhook
        db = current_app.extensions['sqlalchemy']
        me = f"{request.user.username}@{DOMAIN}"

        sent_to = set(r[0] for r in
            db.session.query(Message.receiver).filter(Message.sender == me).distinct())
        received_from = set(r[0] for r in
            db.session.query(Message.sender).filter(Message.receiver == me).distinct())

        partners = sent_to | received_from
        chats = []

        for partner in sorted(partners):
            display_name = partner.split("@")[0]
            profile_pic = None
            status = "offline"

            webhook = None
            if partner.endswith(f"@{DOMAIN}") and partner.startswith("webhook-"):
                webhook_id = partner.split("@")[0][8:]  # Remove "webhook-" prefix
                webhook = db.session.query(Webhook).filter_by(id=webhook_id, user_id=request.user.id).first()

            if webhook:
                display_name = webhook.name
                profile_pic = webhook.profile_pic
            else:
                try:
                    profile = get_profile(partner)
                except ConnectionError:
                    profile = None

                if profile:
                    display_name = profile["display_name"]
                    profile_pic = profile["profile_pic"]
                    status = profile["status"]

            unread_count = db.session.query(Message).filter(
                Message.sender == partner,
                Message.receiver == me,
                Message.is_read == False
            ).count()

            last_msg = db.session.query(Message).filter(
                or_(
                    (Message.sender == partner) & (Message.receiver == me),
                    (Message.sender == me) & (Message.receiver == partner)
                )
            ).order_by(Message.timestamp.desc()).first()

            chats.append({
                "id": partner,
                "display_name": display_name,
                "profile_pic": profile_pic,
                "status": status,
                "unread_count": unread_count,
                "last_message_text": last_msg.text if last_msg else None,
                "last_message_sender": last_msg.sender if last_msg else None,
            })

        return chats


# ── /<target>/messages ────────────────────────────────────────────────────

@chats_blp.route("/<path:target>/messages")
class ChatMessagesResource(MethodView):
    @chats_blp.arguments(MessagesQueryArgs, location="query")
    @chats_blp.response(200, MessageObject(many=True))
    def get(self, args, target):
        """Get messages for a conversation"""
        require_auth()
        from app import Message, DOMAIN
        from json_discovery import get_endpoint
        from sqlalchemy import or_
        import requests as http_requests

        db = current_app.extensions['sqlalchemy']
        me = f"{request.user.username}@{DOMAIN}"

        # Channel server (external)
        if "#" in target:
            target_domain = target.split("#")[0]
            url = get_endpoint(target_domain, "channelserver", "channel_poll")
            if not url:
                url = f"http://{target_domain}/api/channel/poll"
            params = {"path": target, "limit": args["limit"], "since": args.get("since"), "before": args.get("before")}
            try:
                return http_requests.get(url, params=params).json()
            except Exception:
                return []

        target_full = f"{target}@{DOMAIN}" if "@" not in target else target

        sent = (Message.sender == me) & (Message.receiver == target_full)
        received = (Message.sender == target_full) & (Message.receiver == me)
        query = db.session.query(Message).filter(or_(sent, received))

        if args.get("since"):
            query = query.filter(Message.timestamp > datetime.fromtimestamp(args["since"]))
        if args.get("before"):
            query = query.filter(Message.timestamp < datetime.fromtimestamp(args["before"]))

        msgs = query.order_by(Message.timestamp.desc()).limit(args["limit"]).all()

        if target_full and target_full != me:
            unread_msgs = db.session.query(Message).filter(
                Message.sender == target_full,
                Message.receiver == me,
                Message.is_read == False
            ).all()
            if unread_msgs:
                for m in unread_msgs:
                    m.is_read = True
                db.session.commit()

        return [_serialize_message(m) for m in reversed(msgs)]

    @chats_blp.arguments(SendMessageBody)
    @chats_blp.response(201, MessageObject)
    def post(self, data, target):
        """Send a message to a conversation"""
        require_auth()
        from app import Message as MsgModel, DOMAIN
        from json_discovery import get_endpoint
        import requests as http_requests

        db = current_app.extensions['sqlalchemy']

        receiver = f"{target}@{DOMAIN}" if "@" not in target else target
        sender = f"{request.user.username}@{DOMAIN}"

        # Block sending to webhooks (one-way only)
        username = receiver.split("@")[0]
        if username.startswith("webhook-"):
            abort(403, message="Cannot send messages to webhooks")

        # Verify the recipient exists before saving
        if "#" not in target:
            try:
                profile = get_profile(receiver)
            except ConnectionError:
                abort(502, message="Failed to reach remote server")
            if not profile:
                abort(404, message="User not found")

        msg_uuid = str(uuid.uuid4())
        full_id = f"{DOMAIN}/{msg_uuid}"
        val_key = "key-" + msg_uuid[:8]

        new_msg = MsgModel(
            id=full_id,
            sender=sender,
            receiver=receiver,
            text=data["text"],
            validation_key=val_key,
        )
        db.session.add(new_msg)
        db.session.commit()

        if receiver.endswith(f"@{DOMAIN}"):
            from app import User, send_push_notification
            recipient = db.session.query(User).filter_by(username=receiver.split("@", 1)[0]).first()
            if recipient and recipient.id != request.user.id:
                send_push_notification(
                    recipient,
                    f"New message from {request.user.username}",
                    data["text"],
                    url='/',
                )

        # Federate
        msg_payload = {
            "id": full_id,
            "sender": sender,
            "receiver": receiver,
            "text": data["text"],
            "validationKey": val_key,
        }

        if "#" in target:
            target_domain = target.split("#")[0]
            url = get_endpoint(target_domain, "channelserver", "channel_send")
            if not url:
                url = f"http://{target_domain}/api/channel/send"
            try:
                http_requests.post(url, json=msg_payload, timeout=3)
            except Exception:
                pass
        else:
            target_domain = receiver.split("@")[-1]
            url = get_endpoint(target_domain, "userserver", "federation_receive")
            if not url:
                url = f"http://{target_domain}/federation/receive"
            try:
                http_requests.post(url, json=msg_payload, timeout=5)
            except Exception:
                pass

        return _serialize_message(new_msg)


def _serialize_message(msg):
    """Convert a Message model to a response dict."""
    return {
        "id": msg.id,
        "sender": msg.sender,
        "receiver": msg.receiver,
        "text": msg.text,
        "timestamp": msg.timestamp.timestamp(),
        "is_read": msg.is_read,
    }

File: crates/userserver/src/routes/chats.rs
//! `/api/chats` — conversation list, message history, sending, deletion.

use crate::auth::AuthUser;
use crate::profile::get_profile;
use crate::state::AppState;
use axum::extract::{Path, Query, State};
use axum::http::StatusCode;
use axum::routing::get;
use axum::{Json, Router};
use bscp_common::federation::{self, FedMessage};
use bscp_common::models::{Message, User, Webhook};
use bscp_common::{now_ts, uuid, ApiError};
use serde::Deserialize;
use serde_json::{json, Value};
use std::collections::BTreeSet;

pub fn router() -> Router<AppState> {
    Router::new()
        .route("/api/chats", get(chat_list))
        .route("/api/chats/", get(chat_list))
        .route("/api/chats/:target/messages", get(get_messages).post(send_message))
        .route("/api/chats/:target/messages/:message_id", axum::routing::delete(delete_message))
}

fn serialize_message(m: &Message) -> Value {
    json!({
        "id": m.id,
        "sender": m.sender,
        "receiver": m.receiver,
        "text": m.text,
        "timestamp": m.timestamp,
        "is_read": m.is_read,
    })
}

// ── GET /api/chats ───────────────────────────────────────────────────────

async fn chat_list(State(state): State<AppState>, auth: AuthUser) -> Result<Json<Value>, ApiError> {
    let me = state.full_id(&auth.user.username);
    let domain_suffix = format!("@{}", state.domain());

    let sent: Vec<String> =
        sqlx::query_scalar("SELECT DISTINCT receiver FROM messages WHERE sender = ?")
            .bind(&me)
            .fetch_all(&state.pool)
            .await?;
    let recv: Vec<String> =
        sqlx::query_scalar("SELECT DISTINCT sender FROM messages WHERE receiver = ?")
            .bind(&me)
            .fetch_all(&state.pool)
            .await?;

    let partners: BTreeSet<String> = sent.into_iter().chain(recv).collect();
    let mut chats = Vec::with_capacity(partners.len());

    for partner in partners {
        let mut display_name = partner.split('@').next().unwrap_or(&partner).to_string();
        let mut profile_pic: Option<String> = None;
        let mut status = "offline".to_string();

        let webhook_local = partner
            .strip_suffix(&domain_suffix)
            .and_then(|local| local.strip_prefix("webhook-"));

        if let Some(webhook_id) = webhook_local {
            let w = sqlx::query_as::<_, Webhook>("SELECT * FROM webhooks WHERE id = ? AND user_id = ?")
                .bind(webhook_id)
                .bind(&auth.user.id)
                .fetch_optional(&state.pool)
                .await?;
            if let Some(w) = w {
                display_name = w.name;
                profile_pic = w.profile_pic;
            }
        } else if let Ok(Some(profile)) = get_profile(&state, &partner).await {
            if let Some(v) = profile.get("display_name").and_then(|v| v.as_str()) {
                display_name = v.to_string();
            }
            profile_pic = profile.get("profile_pic").and_then(|v| v.as_str()).map(String::from);
            if let Some(v) = profile.get("status").and_then(|v| v.as_str()) {
                status = v.to_string();
            }
        }

        let unread: i64 = sqlx::query_scalar(
            "SELECT COUNT(*) FROM messages WHERE sender = ? AND receiver = ? AND is_read = 0",
        )
        .bind(&partner)
        .bind(&me)
        .fetch_one(&state.pool)
        .await?;

        let last: Option<(String, String)> = sqlx::query_as(
            "SELECT text, sender FROM messages \
             WHERE (sender = ? AND receiver = ?) OR (sender = ? AND receiver = ?) \
             ORDER BY timestamp DESC LIMIT 1",
        )
        .bind(&partner)
        .bind(&me)
        .bind(&me)
        .bind(&partner)
        .fetch_optional(&state.pool)
        .await?;

        chats.push(json!({
            "id": partner,
            "display_name": display_name,
            "profile_pic": profile_pic,
            "status": status,
            "unread_count": unread,
            "last_message_text": last.as_ref().map(|(t, _)| t.clone()),
            "last_message_sender": last.as_ref().map(|(_, s)| s.clone()),
        }));
    }

    Ok(Json(json!(chats)))
}

// ── GET /api/chats/:target/messages ─────────────────────────────────────

#[derive(Deserialize)]
struct MessagesQuery {
    since: Option<f64>,
    before: Option<f64>,
    #[serde(default = "default_limit")]
    limit: i64,
}
fn default_limit() -> i64 {
    50
}

async fn get_messages(
    State(state): State<AppState>,
    auth: AuthUser,
    Path(target): Path<String>,
    Query(q): Query<MessagesQuery>,
) -> Result<Json<Value>, ApiError> {
    let me = state.full_id(&auth.user.username);

    if target.contains('#') {
        let target_domain = target.split('#').next().unwrap_or_default();
        let data = federation::poll_channel(&state.discovery, target_domain, &target, q.limit, q.since, q.before).await;
        return Ok(Json(data));
    }

    let target_full = if target.contains('@') { target.clone() } else { format!("{target}@{}", state.domain()) };

    let mut sql = String::from(
        "SELECT * FROM messages WHERE ((sender = ? AND receiver = ?) OR (sender = ? AND receiver = ?))",
    );
    if q.since.is_some() {
        sql.push_str(" AND timestamp > ?");
    }
    if q.before.is_some() {
        sql.push_str(" AND timestamp < ?");
    }
    sql.push_str(" ORDER BY timestamp DESC LIMIT ?");

    let mut query = sqlx::query_as::<_, Message>(&sql)
        .bind(&me)
        .bind(&target_full)
        .bind(&target_full)
        .bind(&me);
    if let Some(s) = q.since {
        query = query.bind(s);
    }
    if let Some(b) = q.before {
        query = query.bind(b);
    }
    query = query.bind(q.limit.clamp(1, 500));
    let mut msgs = query.fetch_all(&state.pool).await?;

    if target_full != me {
        sqlx::query("UPDATE messages SET is_read = 1 WHERE sender = ? AND receiver = ? AND is_read = 0")
            .bind(&target_full)
            .bind(&me)
            .execute(&state.pool)
            .await?;
    }

    msgs.reverse();
    Ok(Json(json!(msgs.iter().map(serialize_message).collect::<Vec<_>>())))
}

// ── POST /api/chats/:target/messages ───────────────────────────────────

#[derive(Deserialize)]
struct SendBody {
    text: String,
}

async fn send_message(
    State(state): State<AppState>,
    auth: AuthUser,
    Path(target): Path<String>,
    Json(body): Json<SendBody>,
) -> Result<(StatusCode, Json<Value>), ApiError> {
    let sender = state.full_id(&auth.user.username);
    let receiver = if target.contains('@') { target.clone() } else { format!("{target}@{}", state.domain()) };

    let recv_username = receiver.split('@').next().unwrap_or_default();
    if recv_username.starts_with("webhook-") {
        return Err(ApiError::forbidden("Cannot send messages to webhooks"));
    }

    let is_channel = target.contains('#');
    if !is_channel {
        match get_profile(&state, &receiver).await? {
            Some(_) => {}
            None => return Err(ApiError::not_found("User not found")),
        }
    }

    let msg_uuid = uuid();
    let full_id = format!("{}/{}", state.domain(), msg_uuid);
    let val_key = format!("key-{}", &msg_uuid[..8]);
    let ts = now_ts();

    sqlx::query(
        "INSERT INTO messages (id, sender, receiver, text, validation_key, timestamp, is_read) \
         VALUES (?, ?, ?, ?, ?, ?, 0)",
    )
    .bind(&full_id)
    .bind(&sender)
    .bind(&receiver)
    .bind(&body.text)
    .bind(&val_key)
    .bind(ts)
    .execute(&state.pool)
    .await?;

    // Local push (background).
    if let Some(local) = receiver.strip_suffix(&format!("@{}", state.domain())) {
        if let Some(recipient) = sqlx::query_as::<_, User>("SELECT * FROM users WHERE username = ?")
            .bind(local)
            .fetch_optional(&state.pool)
            .await?
        {
            if recipient.id != auth.user.id {
                let (pool, vapid, disc) =
                    (state.pool.clone(), state.vapid.clone(), state.discovery.clone());
                let (title, text) = (format!("New message from {}", auth.user.username), body.text.clone());
                tokio::spawn(async move {
                    bscp_common::push::send_to_user(&pool, disc.client(), &vapid, &recipient.id, &title, &text, "/").await;
                });
            }
        }
    }

    // Federate (background, best-effort).
    let payload = FedMessage {
        id: full_id.clone(),
        sender: sender.clone(),
        receiver: receiver.clone(),
        text: body.text.clone(),
        validation_key: val_key.clone(),
    };
    let disc = state.discovery.clone();
    let target_clone = target.clone();
    let receiver_clone = receiver.clone();
    tokio::spawn(async move {
        if let Some(dom) = target_clone.contains('#').then(|| target_clone.split('#').next().unwrap_or_default().to_string()) {
            federation::deliver_channel(&disc, &dom, &payload).await;
        } else if let Some(dom) = receiver_clone.rsplit('@').next() {
            federation::deliver_dm(&disc, dom, &payload).await;
        }
    });

    let msg = Message {
        id: full_id,
        sender,
        receiver,
        text: body.text,
        validation_key: Some(val_key),
        timestamp: ts,
        is_read: false,
    };
    Ok((StatusCode::CREATED, Json(serialize_message(&msg))))
}

// ── DELETE /api/chats/:target/messages/:message_id ─────────────────────

async fn delete_message(
    State(state): State<AppState>,
    auth: AuthUser,
    Path((_target, message_id)): Path<(String, String)>,
) -> Result<StatusCode, ApiError> {
    let me = state.full_id(&auth.user.username);
    let msg = sqlx::query_as::<_, Message>("SELECT * FROM messages WHERE id = ?")
        .bind(&message_id)
        .fetch_optional(&state.pool)
        .await?;
    let Some(msg) = msg else {
        return Err(ApiError::not_found("Message not found"));
    };
    if msg.sender != me {
        return Err(ApiError::forbidden("Cannot delete other user's messages"));
    }
    sqlx::query("DELETE FROM messages WHERE id = ?").bind(&message_id).execute(&state.pool).await?;
    Ok(StatusCode::NO_CONTENT)
}
